src/queries.cpp

```diff
@@ -149,12 +150,10 @@ void perftest(const std::string &index_filename,
     if (thresholds_filename) {
         std::string t;
         std::ifstream tin(*thresholds_filename);
-    size_t idx = 0;
-        while (std::getline(tin, t)) {
-            thresholds[idx] = std::stof(t);
-        idx += 1;
-        }
-        assert(idx == queries.size());
+        auto t_begin = std::istream_iterator<float>(tin);
+        auto t_end = std::istream_iterator<float>();
+        assert(std::distance(t_begin, t_end) == queries.size());
```

Comment:
:+1: for the check but it won't make a difference with `assert` because I doubt you will run it in debug anyway. There's no reason to disable this for release, so I suggest a hard fail here. Maybe we should have some kind of `unconditional_assert` or `release_assert` or whatever better name you can come up with, because I'm expecting this to be in several places at least. I suggest it to simply `spdlog::error(...)` a message and then invoke `std::exit(1)`.

I used something similar but throwing error instead. But it's not a bad idea to have a version that exits right away, too.

```cpp
template <typename Message>
inline void runtime_assert(bool condition, Message&& message)
{
    if (not condition) {
        if constexpr (std::is_invocable_r_v<std::string, Message>) {
            throw std::runtime_error(message());
        } else {
            throw std::runtime_error(std::forward<Message>(message));
        }
    }
}
```

It's good to have something that also accepts an invokable so you don't have to evaluate stuff if `condition == true`, say:

```cpp
runtime_assert(some_condition, [&] { return string_builder.str(); });
```
Here, you only pay for the `if` if everything's okay. I learned that from rust where it's a standard lint: https://rust-lang.github.io/rust-clippy/master/#or_fun_call :)